src/tabs/console/ConsoleAutocomplete.tsx

```diff
@@ -0,0 +1,70 @@
+// Tab completion engine for autocompleting commands. Inspired by the TabCompletionEngine class
+// for the Terasology in-game console.
+
+export class ConsoleAutocomplete {
+
+  // autocomplete the command.
+  public static complete(command: string): string {
+    this.previousCommands.push(command);
+    // if command was not entered, display some help text
+    if (command.length <= 0) {
+      return "Type 'help' to list all commands.";
+    }
+    // if no commands were previously entered, find a list of commands
+    if (this.previousCommands.length === 1) {
+      this.findMatches(command);
+    }
+    // if matches were found, do nothing to the command
+    if (this.matches.length === 0) {
+      return command;
+    } else if (this.matches.length === 1) {
+      return this.matches[0];
+    }
+    // If the last possible match has already been iterated through, reset the index to the first location.
+    // The match index must be offset because the length function returns 1 with one element, but its index
```

Comment:
I think this comment (only lines 24-25, 23 is useful) isn't really necessary since it works this way (almost?) everywhere, index of last element is length-1